plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -800,23 +808,25 @@ enterComputeCek = computeCek
         -- "apparently good" value.
         (VConstr i _) | fromIntegral @_ @Integer i > fromIntegral @Int @Integer maxBound ->
                         throwErrorDischarged (StructuralError (MissingCaseBranchMachineError i)) e
-        -- Otherwise, we can safely convert the index to an Int and use it
+        -- Otherwise, we can safely convert the index to an Int and use it.
         (VConstr i args) -> case (V.!?) cs (fromIntegral i) of
             Just t  -> computeCek (transferArgStack args ctx) env t
             Nothing -> throwErrorDischarged (StructuralError $ MissingCaseBranchMachineError i) e
+        -- Proceed with caser when expression given is not Constr.
         VCon val -> case unCaserBuiltin ?cekCaserBuiltin val cs of
-            Left err  -> throwErrorDischarged (OperationalError $ CekCaseBuiltinError err) e
-            Right res -> computeCek ctx env res
+            Left err          -> throwErrorDischarged (OperationalError $ CekCaseBuiltinError err) e
+            Right (HeadOnly fX) -> computeCek ctx env fX
+            Right (HeadSpine f xs) -> computeCek (transferConstantSpine xs ctx) env f
```

Comment:
Using `Either` here was stupid on my part, should've introduced a dedicated strict data type.

But now that I look at it, it's probably best to inline `HeadSpine` business in such a data type, so that it has three constructors (error / success with only head / success with head and spine).

Not in this PR though, let's do it separately and measure whether it helps.

Reply:
I'll make a PR for this after this one